dagql/idtui/golden_test.go

```diff
@@ -0,0 +1,360 @@
+
+func TestMain(m *testing.M) {
+	testCtx = telemetry.InitEmbedded(testCtx, nil)
+	code := m.Run()
+	telemetry.Close()
+	os.Exit(code) // don't use defer!
+}
+
+const InstrumentationLibrary = "dagger.io/golden"
+
+func Tracer() trace.Tracer {
+	return otel.Tracer(InstrumentationLibrary)
+}
+
+func Logger() log.Logger {
+	return telemetry.Logger(testCtx, InstrumentationLibrary)
+}
+
+func Middleware() []testctx.Middleware {
+	return []testctx.Middleware{
+		testctx.WithParallel,
+		testctx.WithOTelLogging(Logger()),
+		testctx.WithOTelTracing(Tracer()),
+	}
+}
+
+type TelemetrySuite struct {
+	Home string
+}
+
+func TestTelemetry(t *testing.T) {
+	testctx.Run(testCtx, t, TelemetrySuite{
+		Home: t.TempDir(),
+	}, Middleware()...)
+}
+
+func (s TelemetrySuite) TestGolden(ctx context.Context, t *testctx.T) {
+	for _, ex := range []Example{
+		{Function: "hello-world"},
+		{Function: "fail-log", Fail: true},
+		{Function: "fail-effect", Fail: true},
+		{Function: "fail-log-native", Fail: true},
+		{Function: "encapsulate"},
+		{Function: "pending", Fail: true},
+		{Function: "use-exec-service"},
+		{Function: "use-no-exec-service"},
+		{Module: "./viztest/broken", Function: "broken", Fail: true},
+	} {
+		t.Run(ex.Function, func(ctx context.Context, t *testctx.T) {
+			out, _ := ex.Run(ctx, t, s)
+			golden.Assert(t, out, t.Name())
+		})
+	}
+}
+
+type Example struct {
+	Module   string
+	Function string
+	Args     []string
+	Fail     bool
+}
+
+func (ex Example) Run(ctx context.Context, t *testctx.T, s TelemetrySuite) (string, *dagui.DB) {
+	db, otlpL := testDB(t)
+
+	if ex.Module == "" {
+		ex.Module = "./viztest"
+	}
+
+	daggerBin := "dagger" // $PATH
+	if bin := os.Getenv("_EXPERIMENTAL_DAGGER_CLI_BIN"); bin != "" {
+		daggerBin = bin
+	}
+
+	daggerArgs := []string{"--progress=report", "call", "-m", ex.Module, ex.Function}
+	daggerArgs = append(daggerArgs, ex.Args...)
+
+	// NOTE: we care about CACHED states for these tests, so we need some way for
+	// them to not be flaky (cache hit/miss), but still produce the same golden
+	// output every time. So, we run everything twice. The first run will cache
+	// the things that should be cacheable, and the second run will be the final
+	// result. Each test is responsible for busting its own caches.
+	func() {
+		ctx, span := Tracer().Start(ctx, "warmup")
+		defer telemetry.End(span, func() error { return nil })
+		warmup := exec.Command(daggerBin, daggerArgs...)
+		warmup.Env = append(
+			os.Environ(),
+			fmt.Sprintf("HOME=%s", s.Home), // ignore any local Dagger Cloud auth
+		)
+		warmup.Env = append(warmup.Env, telemetry.PropagationEnv(ctx)...)
+		err := warmup.Run()
+		if ex.Fail {
+			require.Error(t, err)
+		} else {
+			require.NoError(t, err)
+		}
+		// FIXME: there appears to be some delay before a cache is able to be hit
+		// by a separate session. no clue where this comes from (sorry) but a 10
+		// second wait passed 25 times in a row.
+		time.Sleep(10 * time.Second)
```

Comment:
@sipsma Any idea what might be causing this? tl;dr if you run two `dagger call` commands back to back quickly enough it's very possible for the second call to have cache misses, but if you wait ~10+ seconds you'll get cache hits.

Reply:
No immediate clue, that's very surprising and a great find. I don't see us doing anything asynchronously in our session shutdown: https://github.com/sipsma/dagger/blob/22e6902807be2bcb6f1503d9e1643494d9f578c5/engine/server/session.go#L283-L283

Highly possibly buildkit is throwing some cleanup into an async goroutine, but even then I didn't think there was any part of the caching specifically that would only kick in async like that... 🤔

Can you easily repro this with some `dagger calls` on your host?

Reply:
> Can you easily repro this with some `dagger calls` on your host?

Strangely enough, I can't. 🤔 And if I run these tests outside of Dagger against `./hack/dev` they pass just fine without the 10 second sleep. It _only_ fails when running in Dagger, against an engine running in Dagger.

I'm gonna keep poking at this since it's causing the tests to fail in CI. Will let you know if I find anything.

Reply:
Seems to be related to concurrency, somehow, even with seemingly non-overlapping pipelines.

Fails immediately without the 10 second wait:

```sh
# runs use-exec-service and use-no-exec-service in parallel
#
dagger-dev call test telemetry --run '//service'
```

Passes repeatedly:

```fish
# using a loop instead of --count to ensure a fresh engine each time
while dagger-dev call test telemetry --run '//use-exec-service'; and sleep 1; echo ok: (date); end
```

These two pipelines do not overlap at all - one descends from `redis` and the other descends from `alpine` and `busybox`:


```go
func (*Viztest) NoExecService() *dagger.Service {
	return dag.Container().
		From("redis").
		WithExposedPort(6379). // TODO: would be great to infer this
		AsService()
}

func (v *Viztest) UseNoExecService(ctx context.Context) (string, error) {
	return dag.Container().
		From("redis").
		WithServiceBinding("redis", v.NoExecService()).
		WithEnvVariable("NOW", time.Now().String()).
		WithExec([]string{"redis-cli", "-h", "redis", "ping"}).
		Stdout(ctx)
}
```

```go
func (*Viztest) ExecService() *dagger.Service {
	return dag.Container().
		From("busybox").
		WithExec([]string{"echo", "im cached for good"}).
		WithExec([]string{"echo", "im also cached for good"}).
		WithExec([]string{"echo", "im a buster", time.Now().String()}).
		WithExec([]string{"sleep", "1"}).
		WithExec([]string{"echo", "im busted by that buster"}).
		WithNewFile("/srv/index.html", "<h1>hello, world!</h1>").
		WithExec([]string{"httpd", "-v", "-h", "/srv", "-f"}).
		WithExposedPort(80).
		AsService()
}

func (v *Viztest) UseExecService(ctx context.Context) error {
	_, err := dag.Container().
		From("alpine").
		WithServiceBinding("exec-service", v.ExecService()).
		WithEnvVariable("NOW", time.Now().String()).
		WithExec([]string{"wget", "-q", "-O-", "http://exec-service"}).
		Sync(ctx)
	return err
}
```

Despite that, the issue only happens when they both run. And it fails reliably.

Overall there should be four sessions involved:

* A: use-exec-service warmup
* B: use-no-exec-service warmup
* C: use-exec-service test
* D: use-no-exec-service test

Normally A and B warm the cache so that when C and D run they have a predictable view of things for test assertions. Timing wise the two are within the same ballpark, one doesn't seem decidedly faster or slower than the other.

Still no idea what would cause this to happen, seems like parallel sessions is the only clue. CI has been failing even with the 10 second sleep, so it doesn't seem like I'm out of the woods yet.

Reply:
Seems interesting that I'm able to run these two `pending` examples in parallel just fine, but not in parallel with either of the service examples. Adding a service test to the mix makes both of them see cache misses.

* [✅ pending, ✅ pending-duped](https://dagger.cloud/dagger/traces/3452437bb958e46b07f5fe00b66fda38?span=dd5b48bf55c8e1ff)
* [❌ pending, ❌ pending-duped, ✅ use-no-exec-service](https://dagger.cloud/dagger/traces/57885b5e3bef087ad460f110798730fe?span=c8001d5688e967ba)
* [❌ pending, ❌ pending-duped, ❌ use-exec-service](https://dagger.cloud/dagger/traces/48b05cea960cb42524f49f7957c3a8a0?span=1db347d92155b690)

Maybe something gateway-related?

```go
func (*Viztest) Pending(ctx context.Context) error {
	_, err := dag.Container().
		From("alpine").
		WithExec([]string{"echo", "pending cached for good"}).
		WithExec([]string{"echo", "im also cached for good"}).
		WithEnvVariable("NOW", time.Now().String()).
		WithExec([]string{"sleep", "1"}). // wait a bit to help eyeballing
		WithExec([]string{"false"}).      // fail the pipeline
		WithExec([]string{"sleep", "1"}). // will be pending at the end
		Sync(ctx)
	return err
}

func (*Viztest) PendingDuped(ctx context.Context) error {
	_, err := dag.Container().
		From("alpine").
		WithExec([]string{"echo", "pending duped cached for good"}).
		WithExec([]string{"echo", "im also cached for good"}).
		WithEnvVariable("NOW", time.Now().String()).
		WithExec([]string{"sleep", "1"}). // wait a bit to help eyeballing
		WithExec([]string{"false"}).      // fail the pipeline
		WithExec([]string{"sleep", "1"}). // will be pending at the end
		Sync(ctx)
	return err
}
```

Reply:
For now, I've refactored the `CACHED` state tests out into separate examples (prob the right thing to do anyway) and configured them to just skip if a failure is detected (`Flaky: true`).